text/0000-roadmap-2017.md

```diff
@@ -0,0 +1,577 @@
+[futures library announcement]: http://aturon.github.io/blog/2016/08/11/futures/
+
+### Rust should integrate easily into large build systems
+
+When working with larger organizations interested in using Rust, one of the
+first hurdles we tend to run into is fitting into an existing build
+system. We've been exploring a number of different approaches, each of which
+ends up using Cargo (and sometimes `rustc`) in different ways, with different
+stories about how to incorporate crates from the broader crates.io ecosystem.
+Part of the issue seems to be a perceived overlap between functionality in Cargo
+(and its notion of compilation unit) and in ambient build systems, but we have
+yet to truly get to the bottom of the issues—and it may be that the problem is
+one of communication, rather than of some technical gap.
+
+By the end of 2017, this kind of integration should be *easy*: as a community,
+we should have a strong understanding of best practices, and potentially build
+tooling in support of those practices. And of course, we want to approach this
+goal with Rust's values in mind, ensuring that first-class access to the
+crates.io ecosystem is a cornerstone of our eventual story.
+
+### Rust should integrate easily with C++ code
+
+Rust's current support for interfacing with C is fairly strong, but wrapping a C
+library still involves tedious work mirroring declarations and writing C shims
+or other glue code. Moreover, many projects that are ripe for Rust integration
+are currently using C++, and interfacing with those effectively requires
+maintaining an alternative C wrapper for the C++ APIs. This is a problem both
+for Rust code that wants to employ existing libraries and for those who want to
+integrate Rust into existing C/C++ codebases.
+
+Why C++ rather than the myriad other languages we might focus on? Of any of the
+possible language integrations, C++ came up most consistently across both the
+general survey and the commercial outreach. And that's not too surprising: Rust
+is well-positioned as a C++ replacement, so it's natural for people with
+existing C++ codebases to be looking at Rust but needing a good integration
+story for moving incrementally. In addition, many of the features needed for
+high quality C/C++ binding support are prerequisites for high quality binding
+support in other languages, since runtime internals will generally be accessible
+through a C or C++ interface. So in general, C++ seems like the wisest place to
+focus on the language integration front for the moment.
+
+**The goal should be that using a C++ library in Rust is not much harder than
+using it in C++**. In other words, it should be possible to directly include C++
+headers (e.g., include! {myproject.hpp}) and have the extern declarations, glue
+code, and so forth get generated automatically. This means (eventually) full
+support for interfacing with C++ code that uses features like templates,
+overloading, classes and virtual calls, and so forth.
```

Comment:
> Seems like we addressed...mildly separate things.

@jonathandturner explained why we want better C++ story in our roadmap. You yourself go into details as to how “lets shoot for a better C++ integration” could be implemented. Namely:

> I think this starts just by taking a tool like bindgen and making it drop-dead simple to invoke (include! { ... })

The “why” should probably appear in the RFC text (esp. because its not a goal that has spawned from the survey), and the implementation vector/features/etc should not (even though the language used in the RFC does not commit us to anything).

The best way to resolve my concerns (I’m repeating myself, here) would be reword this section slightly from “The goal should be that using a C++ library in Rust is not much harder than using it in C++” to “FFI to C++ ought to have less friction in Rust!”.

---

As I was writing the last paragraph I realised why exactly that sentence bothers me so much. It’s worded as a goal that’s most likely gonna fail to be achieved; I’m very unlikely to be able to take this document on the 2018 new year eve and tick off this goal as achieved. probably because it feels too ambitious for the time-frame. Aiming for “less friction” makes it much more likely. Everything else being same, the paragraph could look like this instead:

> **Rust should aim to provide a lower friction C++ FFI**. It should be possible to ~~directly include~~ process existing C++ headers (e.g., include! {myproject.hpp}) and have the extern declarations, glue code, and so forth get generated automatically. This means (eventually) ~~full~~ support for interfacing with C++ code that uses features like templates,
overloading, classes and virtual calls, and so forth.